### PR TITLE
Upload Validation Strict Validation

### PR DESCRIPTION
**Pull request by DwayneJengSage**
Reviving and refactoring the Strict Validation code to fit in with our upload validation handler chain. Added unit tests.

If a field has the wrong type or if a required field is missing, this will log a warning. In the future, or for new studies, we could configure this to fail upload validation.

Testing done:
- relevant unit tests
- manually tested upload validation
- ran integration tests to make sure existing Upload behavior wasn't broken

### CODE REVIEW

File: app/org/sagebionetworks/bridge/upload/UploadUtil.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            timestampStr = timestampStr.replaceAll("\\s+", "");
+
+            // Log something, so we can keep track of how often this happens.
+            logger.info("Non-standard timestamp in upload data: " + timestampStr);
+        }
+
+        try {
+            return DateUtils.parseISODateTime(timestampStr);
+        } catch (RuntimeException ex) {
+            return null;
```
Comment (eric-wu) [inline app/org/sagebionetworks/bridge/upload/UploadUtil.java:44]: Should we log a warning here?

File: app/org/sagebionetworks/bridge/upload/StrictValidationHandler.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+                // For attachment types, since they just get exported as raw files, we only need to check if it's
+                // required and present. Specifically, if it's required and it's not present, then that's an error.
+                if (isRequired && !attachmentFieldNameSet.contains(fieldName)) {
+                    errorList.add("Required attachment field " + fieldName + " missing");
+                }
+            } else {
+                JsonNode fieldValueNode = recordDataNode.get(fieldName);
+
+                if (fieldValueNode != null && !fieldValueNode.isNull()) {
+                    boolean isValid = validateJsonField(fieldType, fieldValueNode);
```
Comment (eric-wu) [inline app/org/sagebionetworks/bridge/upload/StrictValidationHandler.java:162]: trivial: Does it justify a boolean variable isValid? It is used only once right after it's initialized.

File: app/org/sagebionetworks/bridge/upload/StrictValidationHandler.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+            case TIMESTAMP:
+                // either it's a string in ISO format, or it's a long in epoch milliseconds
+                if (fieldValueNode.isTextual()) {
+                    DateTime dateTime = UploadUtil.parseIosTimestamp(fieldValueNode.textValue());
+                    return dateTime != null;
+                } else if (fieldValueNode.isIntegralNumber()) {
+                    // any integral value can be converted to an epoch milliseconds, so this is good no
+                    // matter what
+                    return true;
+                } else {
```
Comment (eric-wu) [inline app/org/sagebionetworks/bridge/upload/StrictValidationHandler.java:224]: trivial: this last else block is unnecessary.

File: app/org/sagebionetworks/bridge/upload/UploadUtil.java (hunk 1 above)
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/UploadUtil.java:44]: Added a warning to the catch. Also, updated the other log message to a warning.

File: app/org/sagebionetworks/bridge/upload/StrictValidationHandler.java (hunk 2 above)
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/StrictValidationHandler.java:162]: Functionally, it's all the same in byte code.

Stylistically, I prefer that if there's a function call that does a lot of processing, it shouldn't be mixed with other things like if statements. I'd prefer to leave this as is, unless you have strong feelings about this.

File: app/org/sagebionetworks/bridge/upload/StrictValidationHandler.java (hunk 3 above)
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/upload/StrictValidationHandler.java:224]: Unless the code is identical (like in FLOAT vs INT), I prefer not to use fall-through in switch statements.

Comment (eric-wu) [inline app/org/sagebionetworks/bridge/upload/StrictValidationHandler.java:224]: I mean you will return false but it does not have to be in a else block.

**Pull request merged**